README.md

```diff
@@ -1 +1,67 @@
-# swiftdata
\ No newline at end of file
+# Batching
+
+Deep learning uses a technique called “batching” to better utilize hardware
+resources by processing many data samples as one large tensor:
+
+- Initiating data transfer to an accelerator can be a costly synchronization
+  point; fewer synchronizations lead to higher performance.
+  
+- Accelerators typically produce performance wins by repeating the same
+  calculation over and over, in parallel, without interruption.  Computations on
+  large tensors are easily broken down into similar pieces that can be handled
+  at once.
+  
+- When training, updating weights based on gradients calculated from every
+  sample becomes another bottleneck that prevents large quantities of
+  data from being handled at once.  By computing the average gradient over
+  larger batches and updating weights once for the whole batch, we can reduce
+  the number and frequency of weight updates.
+
+## Terminology
+
+Per https://machinelearningmastery.com/difference-between-a-batch-and-an-epoch/:
+
+* Batch Gradient Descent. Batch Size = Size of Training Set
+* Stochastic Gradient Descent. Batch Size = 1
+* Mini-Batch Gradient Descent. 1 < Batch Size < Size of Training Set
+
+## The Problem Space
+
+- datasets can have uniform or non-uniform samples
+
+- A batch is ultimately processed as one tensor (for unsupervised training or
+  inference), or two tensors having the same shape (for supervised training),
```

Comment:
```suggestion
  inference), or two tensors having the same batch size (for supervised training),
```

Reply:
Inputs and targets usually don't have the same shape, just the first dimension. Typical example in image classification:
- batch of images has size 64 x 224 x 224 x 3
- batch of labels has size 64